README.org
* viroCapt

viroCapt is a bioinformatics pipeline and analysis platform for detecting viral insertions in the human genome from viral capture data.

* Installation

viroCapt is implemented as an R package, which contains the pipeline implemented as a Makefile, dedicated R functions from the package to infer, score and rank insertions candidates, and a Shiny webapp for interactive visualization.

** Dependencies

viroCapt depends on standard bioinformatics tools that need to be installed on the system:
- bowtie2
- samtools
- blat

You will also need ~GNU make~ and ~R~ to run the pipeline itself.\\
All of the R package dependencies will be installed alongside the package using ~devtools~.

** Installation 

All the following commands are to be executed in an R session.

Install ~devtools~

#+begin_src R
install.packages("devtools")
#+end_src

If you have a working git installation and R is configured to access internet

#+begin_src R
devtools::install_github("maximewack/viroCapt")
#+end_src

If not, a local installation is also possible.\\
Download the package archive and extract it, then run

#+begin_src R
devtools::install("/path/to/viroCapt")
#+end_src

This will install the R package that will be used by the pipeline, provide the pipeline and reference files, as well as the Shiny webapp.

* Usage

viroCapt contains two parts: a bioinformatics pipeline that runs in batch and produces static results, and an analysis tool that uses some of those results to produce visualizations.

** Pipeline

The pipeline files are copied to the package installation directory.\\
You can find the location of that directory by executing the following command in R after the package is installed

#+begin_src R
system.file(package = "viroCapt")
#+end_src

To run the pipeline, you need the ~makefile~ file, and the ~refs~ folder, containing the reference genomes for all the HPV variants that were used in the design of the probes for HPV capture.\\
It is recommended to copy those files from the package to where you want to use them.

To run the pipeline, the ~makefile~ and ~refs~ folder should be in the same folder, with the ~fastq~ files to analyze in a subdirectory.\\
All ~fastq~ files should match one of these patterns:
- ~xxx.R1.fastq~ and ~xxx.R2.fastq~ for pair-end read sequencing
- ~xxx.U.fastq~ for single read sequencing

The ~fastq~ files can also be gzipped.

You can then run the pipeline with ~make~, with the following options available (and their default value):
- *DATADIR* (fastqs): folder with the ~fastq~ files to analyse
- *THREADS* (2): number of threads to use for ~bowtie2~ and ~blat~
- *QCDIR* (01_QC): destination of the QC plots
- *GLOBALDIR* (02_Global): destination of the global alignement and crude genotyping results
- *LOCALDIR* (03_Local): destination of the local alignement files and genotyping plots
- *FASTADIR* (04_Fasta): destination of the ~fasta~ files of the candidate human sequences
- *BLATDIR* (05_Blat): destination of the raw ~blat~ results, cleaned results and ranked results
- *VISUDIR* (06_Visu): destination of the files used by the analysis tool
- *BT_OPTIONS_GLOBAL* (--no-unal --very-sensitive): ~bowtie2~ options for global alignment
- *BT_OPTIONS_LOCAL* (--local --no-unal --very-sensitive-local): ~bowtie2~ options for local alignement
- *BLAT_OPTIONS* (-minScore=25 -minIdentitiy=90 -noHead): ~blat~ options for the human alignment
- *LOCAL_PLOT_LIMIT* (5): top n genotypes to show in the local plots
- *FINAL_PLOT_LIMIT* (1): top n genotypes to show in the final plots

For example, if your ~fastq~ files reside in the ~fastq_samples~ folder, and you want to run the pipeline on 8 CPU threads:

#+begin_src sh
make DATADIR=fastq_samples THREADS=8
#+end_src

The default ~make~ target is *All*, which executes the *QC*, *Local_plot*, *Final_plot*, and *Visu* targets.\\
You specify any of those targets to re-run the pipeline up to that target.\\
For example, if you want to change the number of genotypes presented in some local plots, you can delete the plots that you want re-generated and run

#+begin_src sh
make DATADIR=fastq_samples LOCAL_PLOT_LIMIT=8 Local_plot
#+end_src

The pipeline will print the output of every command as it is used, and will print execution milestones in red.\\
All the analysis results will appear in the ~01_*~ to ~06_*~ folders within the folder containing the source ~fastq~ files.

The bowtie index files for all references will be built as needed, as well as the download of the reference human genome from the UCSC, usually with the first run.

** Visualization

Once the pipeline has successfully run, you can use the ~rds~ files (in the ~06_Visu~ folder) in the visualization tool.

In R, run

#+begin_src R
library(viroCapt)
visu()
#+end_src

or simply

#+begin_src R
viroCapt::visu()
#+end_src

You can use any option of ~shiny::runApp()~ in ~viroCapt::visu()~ to customize how the Shiny app runs.\\
For example, if you want to make the tool available to other users in the same network on port 1234

#+begin_src R
viroCapt::visu(host = "0.0.0.0",
               port = 1234)
#+end_src

* Test data

The ~fastq~ files produced by running HPV capture on the *HeLa cell line* are distributed with the package, in the ~hela/~ folder in the files installed by the package.\\
They are accompanied by the expected results files and can be used to check the package is correctly installed and running.



NEWS.org
* 3.2.0
  
** New features

- compute a majority consensus sequence in the same loop and output it as a fasta file in $LOCALDIR

** Changes

- parallelize CIGAR parsing
- run depth profile computation as a loop (should use less memory)
- keep the computed local profile

** Bugfix

- rename HPV58_HPV58sc***.fasta files to conform to the standard naming pattern

* 3.1.0

- add a NEWS.org file to track changes between versions
- add a README.org file to explain how to install and use viroCapt
- allow `visu()` to pass args to `shiny::runApp`


R/utils.R
#' Compose two functions (sequence order, not composition order)
#'
#' @param f A function a -> b
#' @param g A function b -> c
#' @return A function a -> c
compose2 <- function(f, g)
{
  function(...)
  {
    g(f(...))
  }
}


#' Compose a list of functions
#'
#' @param funs A list of functions, in sequence order
#' @return A composed function
compose <- function(funs)
{
  Reduce(compose2, funs)
}


#' Compose all combinations of the given named functions (left to right order)
#'
#' @param ... Named functions to compose-combine
#' @return A list of functions, result of the compositions of all ordered combinations of inputted functions
compose_combn <- function(...)
{
  funs <- lapply(list(...), match.fun)

  fun_names <- names(funs)
  sizes <- length(funs)

  1:sizes %>%
  lapply(utils::combn, x = fun_names) %>%
    lapply(apply, 2, function(x) {paste0(x, collapse = "")}) %>%
    unlist -> comb_names

  1:sizes %>%
    lapply(utils::combn, x = funs) %>%
    lapply(apply, 2, compose) %>%
    unlist %>%
    stats::setNames(comb_names)
}


R/pipeline.R
#' Plot depth
#'
#' @export
#' @param stem Stem name for sam file and plot file
#' @param limit Limit the number of genotypes to display
plot_depth <- function(stem, limit = 5)
{
  paste0(stem, ".rds") %>%
    readRDS -> sam

    sam %>%
      limit_genotypes(limit) %>%
      ggplot_depth -> Plot

    paste0(stem, ".png") %>%
      ggplot2::ggsave(Plot)
}


#' Create the profile object
#'
#' @export
#' @param stem Stem name of the sam file
create_profile <- function(stem, threads, consensus = F)
{
  paste0(stem, ".sam") %>%
    read_sam -> sam

  if (length(sam) > 0)
  {
    sam %>%
      dplyr::mutate(parsed = cigar %>% parse_cigar(threads)) %>%
      tidyr::unnest(parsed) %>%
      read_depth(consensus = consensus) %>%
      saveRDS(paste0(stem, ".rds"))
  } else
  {
    data.frame(genotype = "No result", pos = 0, n = 0, stringsAsFactors = T) %>%
      saveRDS(paste0(stem, ".rds"))
  }
}

#' Create the visualisation object
#'
#' @export
#' @param sam Stem name of the rds file for the sequencing depth
#' @param summ Stem name of the blat summary file
#' @param visu Name of the RDS file to same the viz data to
create_visu <- function(sam, summ, visu)
{
  paste0(summ, "_summary.tsv") %>%
    read_summary -> summ_blat

  paste0(sam, ".rds") %>%
    readRDS -> sam

  list(summary = summ_blat,
       profile = sam) %>%
  saveRDS(visu)
}


#' Plot the final plot
#'
#' @export
#' @param sam Stem name of the sam file
#' @param summ Stem name of the blat summary file
#' @param limit Limit the number of genotypes to display
plot_final <- function(sam, summ, limit = 1)
{
  paste0(summ, "_summary.tsv") %>%
    read_summary -> summ_blat

  paste0(sam, ".rds") %>%
    readRDS -> sam

  if (length(summ_blat) > 0)
  {
    sam %>%
      limit_genotypes(limit) -> sam

    summ_blat %>%
      dplyr::semi_join(sam, by = "genotype") -> summ_blat

    sam %>%
      ggplot_final(summ_blat) -> Plot

    paste0(summ, ".png") %>%
      ggplot2::ggsave(Plot)
  } else
  {
    cat(NULL, file = paste0(summ, ".png"))
  }
}


#' Extract non-viral sequences as fasta files
#'
#' @export
#' @param samfile A local alignment sam file
#' @param fastafile The fasta file to write the result to
extract_fasta <- function(samfile, fastafile, threads)
{
  samfile %>%
    read_sam -> sam

  if (length(sam) > 0)
  {
    sam %>%
      dplyr::filter(grepl("S", cigar)) %>%
      dplyr::mutate(parsed = cigar %>% parse_cigar(threads),
                    parsed = purrr::map2(parsed, pos, extract_features)) %>%
      tidyr::unnest(parsed) %>%
      dplyr::filter(length_read > 25,
                    feature %in% c("left", "right")) %>%
      extract_unaligned %>%
      write_fasta(fastafile)
  } else
  {
    cat(NULL, file = fastafile)
  }
}


#' Clean blat file
#'
#' @export
#' @param blatfile A headerless blat file name
#' @param fastafile A headerless fasta file name
#' @param cleaned A file name for output
cleaned_blat <- function(blatfile, fastafile, cleaned)
{
  fastafile %>%
    read_fasta -> fasta

  blatfile %>%
    read_blat -> blat

  if (length(blat) > 0)
  {
    blat %>%
      clean_blat %>%
      tag_blat(fasta) %>%
      write_blat(cleaned)
  } else
  {
    cat(NULL, file = cleaned)
  }
}


#' Summarise blat file
#'
#' @export
#' @param blatfile A headerless cleaned blat file name
#' @param summarised A file name for output
summarised_blat <- function(blatfile, summarised)
{
  blatfile %>%
    utils::read.delim() -> blat

  if (length(blat) > 0)
  {
    blat %>%
      summarise_blat %>%
      write_blat(summarised)
  } else
  {
    cat(NULL, file = summarised)
  }
}


inst/visu/ui.R
library(shiny)

ui <- bootstrapPage(
  sidebarLayout(
    sidebarPanel(
      titlePanel("Viral integration pipeline results"),
      inputPanel(fileInput("visu", "Visualisation file", accept = "*.rds")),
      inputPanel(selectizeInput("genotype",
                                "Genotypes",
                                "",
                                multiple = T),
                 selectizeInput("scores",
                                "Scores",
                                "",
                                multiple = T),
                 selectizeInput("chrs",
                                "Chromosomes",
                                "",
                                multiple = T),
                 sliderInput("nreads", "Minimum number of reads", 0, 0, 0, step = 1, round = T),
                 sliderInput("match", "Minimum match score", 0, 0, 0, step = 1, round = T)),
      width = 3),
    mainPanel(plotly::plotlyOutput("plot"),
              DT::dataTableOutput("table"),
              width = 9)))


inst/visu/server.R
library(viroCapt)
library(ggplot2)
library(shiny)

options(shiny.maxRequestSize = 1024 ^ 3)

UIlabels <- function(variable)
{
  variable %>%
    forcats::fct_infreq() %>%
    summary -> varsumm

  names(varsumm) %>%
    stats::setNames(paste0(names(varsumm), " (", varsumm, ")"))
}

server <- function(input, output, session)
{
  # Read the visualisation object

  visu <- reactive({
    req(input$visu)

    readRDS(input$visu$datapath)
  })

  # Plot

  ## Extract the sequencing depth object
  sam <- reactive({
    req(visu())

    visu()$profile
  })

  ## Downsample the depth information
  downsampled_depths <- reactive({
    sam() %>% viroCapt:::downsample(10)
  })

  ## Filter the depths for the selected genotypes
  depths <- reactive({
    req(input$genotype)

    downsampled_depths() %>%
      dplyr::filter(genotype %in% input$genotype)
  })

  ## Base depth plot
  localplot <- reactive({
    viroCapt:::ggplot_depth(depths())
  })

  ## Output the annotated plot
  output$plot <- plotly::renderPlotly({
    max(depths()$n) -> max_n

    if (input$scores %>% is.null)
      localplot() -> p
    else {
      localplot() +
        geom_vline(data = summ_blat(), aes(xintercept = position, color = chr, alpha = quality)) -> p

      if (summ_blat() %>% dplyr::filter(feature == "left") %>% nrow > 0)
        p + geom_segment(data = summ_blat() %>% dplyr::filter(feature == "left"), aes(x = position, xend = position - 100, y = -max_n/20, yend = -max_n/20, color = chr, alpha = quality)) -> p

      if (summ_blat() %>% dplyr::filter(feature == "right") %>% nrow > 0)
        p + geom_segment(data = summ_blat() %>% dplyr::filter(feature == "right"), aes(x = position, xend = position + 100, y = -max_n/20, yend = -max_n/20, color = chr, alpha = quality)) -> p
    }

    p +
      scale_alpha_ordinal(range = c(0.5, 1)) +
      theme_classic()
  })

  # Table

  ## Extract the blat summary
  summ_blat_file <- reactive({
    req(visu())

    visu()$summary %>%
      dplyr::mutate(quality = quality %>% forcats::fct_drop())
  })

  ## Filter the blat summary results on genotype, chr, nreads, score, and match
  summ_blat <- reactive({
    summ_blat_file() %>%
      dplyr::filter(genotype %in% input$genotype,
                    n >= input$nreads,
                    match >= input$match) -> summ

    if (! input$chrs %>% is.null)
      summ %>% dplyr::filter(chr %in% input$chrs) -> summ

    if (! input$scores %>% is.null)
      summ %>% dplyr::filter(quality %in% input$scores) -> summ

    summ
  })

  ## Output the table
  output$table <- DT::renderDataTable(
  {
    summ_blat() %>%
      dplyr::mutate_at(vars(feature, chr, quality), factor)
  },
  options = list(dom = "Bfrtip",
                 buttons = c("copy", "excel"),
                 paging = T,
                 info = F,
                 fixedHeader = T),
  filter = "top",
  extensions = "Buttons")

  # Update UI values

  observe({
    sam() %>%
      dplyr::group_by(genotype) %>%
      dplyr::summarise(n = max(n)) %>%
      dplyr::mutate(genotype = genotype %>% as.character) -> genotypes

    genotypes$genotype %>%
      stats::setNames(paste0(genotypes$genotype, " (", genotypes$n, ")")) -> genotypes

    updateSelectizeInput(session, "genotype", choices = genotypes, selected = genotypes[1])
  })

  observe({
    updateSelectizeInput(session, "scores", choices = summ_blat_file()$quality %>% UIlabels)
    updateSelectizeInput(session, "chrs", choices = summ_blat_file()$chr %>% UIlabels)
    updateSliderInput(session, "nreads", min = summ_blat_file()$n %>% min, max = summ_blat_file()$n %>% max)
    updateSliderInput(session, "match", min = summ_blat_file()$match %>% min, max = summ_blat_file()$match %>% max)
  })

}


R/generator_functions.R

#' Generate a final plot function
#'
#' @export
#' @param write Write plot on disk?
#' @param fastqdir Directory of fastq files
#' @param dir Directory in which to write the plot if write == T
#' @return A function accepting a local plot object, a blat summary, and a fastq name, and returning a plot object
Gplot_final <- function(write = F, fastqdir = NULL, dir = NULL)
{
  if (write)
    dir.create(paste0(fastqdir, "/", dir), showWarnings = F, recursive = T)

  function(local_plot, summ_blat, fastq)
  {
    summ_blat %>% filter(!is.na(quality), quality >= "HC") -> summ_blat

    local_plot$data$n %>% max -> max_n
    local_plot +
      ggplot2::geom_vline(data = summ_blat, ggplot2::aes(xintercept = position, color = chr, alpha = quality)) -> p

    if (summ_blat %>% filter(feature == "left") %>% nrow > 0)
      p + ggplot2::geom_segment(data = summ_blat %>% filter(feature == "left"), ggplot2::aes(x = position, xend = position - 100, y = -max_n/20, yend = -max_n/20, color = chr, alpha = quality)) -> p

    if (summ_blat %>% filter(feature == "right") %>% nrow > 0)
      p + ggplot2::geom_segment(data = summ_blat %>% filter(feature == "right"), ggplot2::aes(x = position, xend = position + 100, y = -max_n/20, yend = -max_n/20, color = chr, alpha = quality)) -> p

    p +
      ggplot2::scale_alpha_ordinal(limits = c("HC", "HCT"), range = c(0.5, 1)) +
      ggplot2::theme_classic() -> Plot

    if (write)
      ggplot2::ggsave(plot = Plot, filename = paste0(fastqdir, "/", dir, "/", fastq, ".png"))
  }
}


R/functions.R
#' Read a headerless sam file
#'
#' @param filename The SAM filename
#' @return A dataframe with columns (read, genotype, pos, cigar, seq)
read_sam <- function(filename)
{
  suppressWarnings({
  readr::read_tsv(filename,
           col_names = c("read", "flag", "genotype", "pos", "mapq", "cigar", "rnext", "pnext", "tlen", "seq"),
           col_types = "ciciicciic") -> sam

  if (length(sam) > 0)
    sam %>%
      dplyr::select(read, genotype, pos, cigar, seq) %>%
      dplyr::mutate(genotype = forcats::fct_infreq(genotype))
  })
}


#' Compute the sequencing depth and create a consensus
#'
#' Tally the nucleotides for all positions
#'
#' @param sam sam file object (dataframe)
#' @param consensus whether to compute consensus sequences or not
#' @return A sequencing depth object
read_depth <- function(sam, consensus)
{
  sam %>%
    dplyr::filter(type %in% c("M", "I")) %>%
    dplyr::mutate(offset = stringr::str_match(cigar, "^(\\d+)S")[,2] %>% as.numeric,
                  offset = ifelse(offset %>% is.na, 0, offset),
                  start_read = start_read - offset,
                  end_read = end_read - offset,
                  start = pos + start_read - 1,
                  end = pos + end_read - 1,
                  subseq = substring(seq, start_read, end_read) %>% stringr::str_split("")) %>%
    dplyr::select(genotype, subseq, start, end) -> sam

  sam %>%
    by(sam$genotype,
       simplify = F,
       function(df)
       {
         size <- max(df$end)
         depth <- if (consensus) {data.frame(pos = 1:size, n = 0, A = 0, T = 0, G = 0, C = 0, consensus = "")
         } else {data.frame(pos = 1:size, n = 0)}

         purrr::pwalk(list(df$start, df$end, df$subseq), function(start, end, subseq)
         {
           depth$n[start:end] <<- depth$n[start:end] + 1
           if (consensus)
           {
             depth$A[start:end] <<- depth$A[start:end] + (subseq == "A")
             depth$T[start:end] <<- depth$T[start:end] + (subseq == "T")
             depth$G[start:end] <<- depth$G[start:end] + (subseq == "G")
             depth$C[start:end] <<- depth$C[start:end] + (subseq == "C")
           }
         })

         if (consensus)
           mapply(function(n, a, t, g, c)
           {
             if (n == 0) "N"
             else names(which.max(c(A = a, T = t, G = g, C = c)))
           }, depth$n, depth$A, depth$T, depth$G, depth$C) -> depth$consensus

         depth
       }) -> depths

  data.frame(genotype = names(depths), data = depths %>% unclass) %>%
    tidyr::unnest(cols = data) -> depths

  depths %>%
    dplyr::group_by(genotype) %>%
    dplyr::summarise(avg_depth = mean(n)) %>%
    dplyr::arrange(desc(avg_depth)) %>%
    dplyr::pull(genotype) -> genotype_order

  depths %>%
    dplyr::mutate(genotype = genotype %>% factor(levels = genotype_order))
}


#' Normalise a sequencing depth object using a reference
#'
#' @param depth A sequencing depth object produced by read_depth
#' @param qc_norm A normalization map for the depth object
#' @return A normalized sequencing depth object
normalise_depth <- function(depth, qc_norm)
{
  depth %>%
    dplyr::left_join(qc_norm, by = c("genotype", "pos")) %>%
    dplyr::mutate(n = n.x / n.y) %>%
    dplyr::select(-n.x, -n.y)
}


#' Downsample and object of sequencing depth
#'
#' The sequencing depth object has a width corresponding to the length of the aligned genome (eg. ~8000bp for HPV)
#' This is way too large for plotting on a regular screen
#' This function allows the downsampling of a depth object to speed up other operations, such as plotting
#'
#' @param depth A sequencing depth object, produced by read_depth
#' @param ratio The downsampling ratio, 5 by default
#' @return A downsampled sequencing depth object
downsample <- function(depth, ratio = 5)
{
  depth %>%
    dplyr::mutate(pos = ( (pos / ratio) %>% ceiling) * ratio) %>%
    dplyr::group_by(genotype, pos) %>%
    dplyr::summarise(n = mean(n, na.rm = T)) %>%
    dplyr::ungroup()
}


#' Compute moving average
#'
#' @param depths A nucleotide depth object
#' @param window Half-size of the window
#' @return A smoothed depth object
MA <- function(depths, window)
{
  depths %>%
    dplyr::filter(n > 0) %>%
    split(.$genotype) %>%
    lapply(function(x)
           {
             x %>%
               dplyr::mutate(roll = pos %>% purrr::map_dbl(~mean(x$n[x$pos %in% seq(. - window, . + window)], na.rm = T))) %>%
               tidyr::complete(pos = 1:max(pos, na.rm = T), fill = list(genotype = x$genotype %>% unique, roll = 0, n = 0))
           }) %>%
    dplyr::bind_rows() %>%
    dplyr::mutate(n = roll)
}


#' Limit the number of genotypes in a sample
#'
#' @param depths A nucleotide depth object
#' @param limit The number of genotypes to keep
#' @return A depth object with only the top n genotypes
limit_genotypes <- function(depths, limit)
{
  depths %>%
    dplyr::filter(genotype %in% base::levels(genotype)[1:limit]) %>%
    dplyr::mutate(genotype = genotype %>% forcats::fct_drop())
}


#' Plot the nucleotide depth
#'
#' @param depth A sequencing depth object
#' @return A nucleotide depth plot
ggplot_depth <- function(depth)
{
  depth %>%
    ggplot2::ggplot() +
    ggplot2::aes(x = pos, y = n) +
    ggplot2::geom_area() +
    ggplot2::facet_grid(~genotype) +
    ggplot2::theme(strip.text = ggplot2::element_text(angle = -90))
}


#' Parse a cigar string into a dataframe with type, length, start and end position
#'
#' @param cigar vector of CIGAR string
#' @return parsed CIGAR as a dataframe
parse_cigar <- function(cigar, threads)
{
  if (threads > 1) future::plan("multisession", workers = threads)

  cigar %>%
    gregexpr(pattern = "\\d+(M|I|D|S|N|H|P|=|X)") %>%
    regmatches(x = cigar) %>%
    furrr::future_map(function(x)
    {
      type <- substring(x, nchar(x))
      length <- substring(x, 0, nchar(x) -1) %>% as.numeric

      length_read <- length
      length_read[type == "D"] <- 0

      length_hpv <- length
      length_hpv[type == "I"] <- 0

      start_read <- cumsum(length_read) - length_read + 1
      start_hpv <- cumsum(length_hpv) - length_hpv + 1

      end_read <- cumsum(length_read)
      end_hpv <- cumsum(length_hpv)

      data.frame(length_read = length_read, length_hpv = length_hpv, type = type, start_read = start_read, end_read = end_read, start_hpv = start_hpv, end_hpv = end_hpv, stringsAsFactors = F)
    })
}


#' Extract relevant features from a parsed CIGAR string
#'
#' @param parsed_cigar A parsed CIGAR as a dataframe (as returned by parse_cigar)
#' @param pos The matching position on the viral sequence for that CIGAR string
#' @return A dataframe of features with their position on the viral sequence
extract_features <- function(parsed_cigar, pos)
{
  parsed_cigar$length_read %>% sum -> read_length

  parsed_cigar %>%
    dplyr::mutate(feature = dplyr::case_when(type == "S" & start_read == 1 ~ "left",
                                             type == "S" & end_read == read_length ~ "right",
                                             T ~ type),
                  feature_pos = dplyr::case_when(feature == "D" ~ pos + start_hpv - 1,
                                                 feature == "I" ~ pos + start_hpv- 2,
                                                 feature == "left" ~ pos + end_hpv,
                                                 feature == "right" ~ pos + start_hpv - 2)) -> df

  if (df$feature[1] == "left")
    df$feature_pos <- df$feature_pos - df$length_read[1]

  df
}


#' Extract unaligned sequences
#'
#' @param parsed A sam object (dataframe) with parsed CIGARs and extracted features
#' @return A sam object with unaligned sequences extracted
extract_unaligned <- function(parsed)
{
  parsed %>%
    dplyr::mutate(nalign_seq = substring(seq, start_read, end_read)) %>%
    dplyr::arrange(feature_pos, dplyr::desc(length_read)) %>%
    dplyr::select(genotype, read, feature, feature_pos, nalign_seq) %>%
    droplevels
}


#' Write a fasta file from a parsed sam file
#'
#' @export
#' @param parsed_sam A parsed sam object (cigar went through parse_cigar and extract_features)
#' @param fastafile A file to write to
write_fasta <- function(parsed_sam, fastafile)
{
  paste(paste0(">",
               parsed_sam$read, "|",
               parsed_sam$genotype, "|",
               parsed_sam$feature, "|",
               parsed_sam$feature_pos),
        parsed_sam$nalign_seq,
        sep = "\n",
        collapse = "\n") %>%
    cat(file = fastafile)
}


#' Read a fasta file
#'
#' @export
#' @param fastafile A headerless fasta file to read
#' @return A fasta file as a dataframe
read_fasta <- function(fastafile)
{
  fastafile %>%
    readLines() -> raw

    data.frame(desc = sub("^>", "", raw[c(T, F)]),
               nalign_seq = raw[c(F, T)],
               stringsAsFactors = F) %>%
      tidyr::separate(desc, into = c("read", "genotype", "feature", "feature_pos"), sep = "\\|") %>%
      dplyr::mutate(feature_pos = feature_pos %>% as.numeric)
}


#' Write a blat file
#'
#' write.table with tab separator and without rownames
#'
#' @export
#' @param x Blat object
#' @param file File name
write_blat <- function(x, file = "")
  utils::write.table(x, file, sep = "\t", row.names = F)


#' Read a headerless blat file
#'
#' @export
#' @param blatfile The file to read
#' @return A blatfile organized as a dataframe
read_blat <- function(blatfile)
{
  blatfile %>%
    readr::read_tsv(col_names = c("match","mis-match","rep.match","N's","Q gap count","Q gap bases","T gap count","T gap bases","strand","Q name","Q size","Q start","Q end","T name","T size","T start","T end","block count","blockSizes","qStarts","tStarts")) -> blat

  if (length(blat) > 0)
    blat %>%
      tidyr::separate(`Q name`, c("read", "genotype", "feature", "position"), sep = "\\|")
}


#' Clean a raw blat file
#'
#' @param blat A blat object (dataframe obtained from Gread_blat)
#' @return A cleaned, deduplicated blat object
clean_blat <- function(blat)
{
  blat %>%
    dplyr::mutate(prop = (`Q end` - `Q start`) / `Q size`,
                  chr_position = dplyr::case_when(feature == "left" & strand == "-" ~ `T start`,
                                                  feature == "left" & strand == "+" ~ `T end`,
                                                  feature == "right" & strand == "+" ~ `T start`,
                                                  feature == "right" & strand == "-" ~ `T end`)) %>%
  dplyr::filter(`Q gap bases` <= 5,
                `T gap bases` <= 5,
                prop > .9) %>%
  dplyr::rename(chr = `T name`, size = `Q size`) %>%
  dplyr::group_by(read, strand, genotype, feature, position, chr, chr_position, size) %>%
  dplyr::summarise(match = max(match, na.rm = T)) %>%
  dplyr::filter(size == max(size, na.rm = T)) %>%
  dplyr::ungroup() %>%
  dplyr::distinct()
}


#' Remove homologs (reads with multiple matches in different places)
#'
#' @param blat A clean blat object (dataframe)
#' @return A filtered blat object with homologs removed
remove_homologs <- function(blat)
{
  blat %>%
    dplyr::group_by(read, genotype, feature) %>%
    dplyr::add_count() %>%
    dplyr::filter(n == 1) %>%
    dplyr::ungroup() %>%
    dplyr::select(-n)
}


#' Find positions with more than one matching read
#'
#' @param blat A clean blat object (dataframe)
#' @return A filtered blat object with only concordant positions kept
find_concordant_positions <- function(blat)
{
  blat %>%
    dplyr::group_by(genotype, feature, position, chr, chr_position) %>%
    dplyr::add_count() %>%
    dplyr::filter(n > 1) %>%
    dplyr::arrange(chr, position, chr_position) %>%
    dplyr::select(-n) %>%
    dplyr::ungroup()
}


#' Find positions with at least one left and right clip on the same chromosome
#'
#' @param blat A clean blat object (dataframe)
#' @return A filtered blat object with only two-sided clips
filter_two_sides <- function(blat)
{
  blat %>%
    dplyr::semi_join(blat %>%
                     dplyr::group_by(genotype, chr) %>%
                     dplyr::summarise(feat = dplyr::n_distinct(feature)) %>%
                     dplyr::filter(feat == 2) %>%
                     dplyr::ungroup())
}


#' Tag a blat object with quality measures
#'
#' @param blat A blat object (dataframe)
#' @param fasta The corresponding fasta file that produced the blat object
#' @return An tagged blat file enriched with the source sequence
tag_blat <- function(blat, fasta)
{
  compose_combn(H = remove_homologs,
                C = find_concordant_positions,
                T = filter_two_sides) -> comb_funs

  lapply(comb_funs, function(x, y) x(y), blat) -> res

  names(res) %>%
    lapply(function(x) {dplyr::mutate(res[[x]], quality = x)}) %>%
    dplyr::bind_rows() %>%
    dplyr::mutate(quality = quality %>% ordered(levels = c("T", "C", "H", "HT", "CT", "HC", "HCT"))) %>%
    dplyr::group_by(read, feature) %>%
    dplyr::filter(quality == max(quality, na.rm = T)) %>%
    dplyr::ungroup() %>%
    dplyr::full_join(blat) %>%
    dplyr::mutate(position = position %>% as.numeric) %>%
    dplyr::left_join(fasta, by = c("genotype", "read", "feature", "position" = "feature_pos")) %>%
    dplyr::arrange(dplyr::desc(quality), chr, position, dplyr::desc(match)) %>%
    dplyr::filter(nchar(nalign_seq) == size) %>%
    dplyr::select(read, strand, sequence = nalign_seq, genotype, feature, position, chr, chr_position, match, quality)
}


#' Summarise the results of a tagged blat file
#'
#' @param blat A tagged blat file
#' @return A summary table of results
summarise_blat <- function(blat)
{
  blat %>%
    dplyr::mutate(quality = quality %>% ordered(levels = c("T", "C", "H", "HT", "CT", "HC", "HCT"))) %>%
    dplyr::group_by(genotype, feature, position, chr, chr_position) %>%
    dplyr::add_count() %>%
    dplyr::summarise(n = max(n, na.rm = T), quality = max(quality, na.rm = T), match = max(match, na.rm = T)) %>%
    dplyr::arrange(dplyr::desc(quality), dplyr::desc(n), dplyr::desc(match)) %>%
    dplyr::ungroup()
}


#' Read a blat summary file
#'
#' @param summ_blat Name of the summary file
#' @return A blat summary object
read_summary <- function(summ_blat)
{
  utils::read.delim(summ_blat) -> blat

  if (length(blat) > 0)
    blat %>%
      dplyr::mutate(quality = quality %>% ordered(levels = c("T", "C", "H", "HT", "CT", "HC", "HCT")))
}


#' Plot the final plot with insertions
#'
#' @param depth A sequencing depth object
#' @param summ_blat A blat summary object
#' @return A nucleotide depth plot
ggplot_final <- function(depth, summ_blat)
{
  depth %>%
    ggplot2::ggplot() +
    ggplot2::aes(x = pos, y = n) +
    ggplot2::geom_area() +
    ggplot2::facet_grid(~genotype) +
    ggplot2::theme(strip.text = ggplot2::element_text(angle = -90)) -> local_plot

    summ_blat %>% dplyr::filter(!is.na(quality), quality >= "HC") -> summ_blat

    max(depth$n) -> max_n

    local_plot +
      ggplot2::geom_vline(data = summ_blat, ggplot2::aes(xintercept = position, color = chr, alpha = quality)) -> p

    if (summ_blat %>% dplyr::filter(feature == "left") %>% nrow > 0)
      p + ggplot2::geom_segment(data = summ_blat %>% dplyr::filter(feature == "left"), ggplot2::aes(x = position, xend = position - 100, y = -max_n/20, yend = -max_n/20, color = chr, alpha = quality)) -> p

    if (summ_blat %>% dplyr::filter(feature == "right") %>% nrow > 0)
      p + ggplot2::geom_segment(data = summ_blat %>% dplyr::filter(feature == "right"), ggplot2::aes(x = position, xend = position + 100, y = -max_n/20, yend = -max_n/20, color = chr, alpha = quality)) -> p

    p +
      ggplot2::scale_alpha_ordinal(limits = c("HC", "HCT"), range = c(0.5, 1)) +
      ggplot2::theme_classic()
}

#' Write the consensus file
#'
#' @export
#' @param stem Stem of the sample
#' @return Nothing, writes the %_consensus.fa file
write_consensus <- function(stem)
{
  readRDS(paste0(stem, ".rds")) -> depth

  depth %>%
    dplyr::group_by(genotype) %>%
    dplyr::do(consensus = paste0(.$consensus, collapse = "")) %>%
    tidyr::unnest(consensus) -> consensus

  paste0(">", consensus$genotype) %>%
    paste(consensus$consensus, sep = "\n") %>%
    cat(file = paste0(stem, "_consensus.fa"), sep = "\n\n")

}
